contributingGuides/STYLE.md

```diff
@@ -511,112 +958,215 @@ In React Native, one **must not** attempt to falsey-check a string for an inline
+export default React.forwardRef(FancyInput)
+```
 
-```javascript
-// Bad
-export default compose(
-    withLocalize,
-)(MyComponent);
+### Hooks and HOCs
+
+Use hooks whenever possible, avoid using HOCs.
+
+> Why? Hooks are easier to use (can be used inside the function component), and don't need nesting or `compose` when exporting the component. It also allows us to remove `compose` completely in some components since it has been bringing up some issues with TypeScript. Read the [`compose` usage](#compose-usage) section for further information about the TypeScript issues with `compose`.
+
+> Note: Because Onyx doesn't provide a hook yet, in a component that accesses Onyx data with `withOnyx` HOC, please make sure that you don't use other HOCs (if applicable) to avoid HOC nesting.
+
+```tsx
+// BAD
+type ComponentOnyxProps = {
+    session: OnyxEntry<Session>;
+};
+
+type ComponentProps = WindowDimensionsProps &
+    WithLocalizeProps &
+    ComponentOnyxProps & {
+        someProp: string;
+    };
+
+function Component({windowWidth, windowHeight, translate, session, someProp}: ComponentProps) {
+    // component's code
+}
 
-// Good
 export default compose(
-    withLocalize,
     withWindowDimensions,
-)(MyComponent);
+    withLocalize,
+    withOnyx<ComponentProps, ComponentOnyxProps>({
+        session: {
+            key: ONYXKEYS.SESSION,
+        },
+    }),
+)(Component);
 
-// Good
-export default withLocalize(MyComponent)
+// GOOD
+type ComponentOnyxProps = {
+    session: OnyxEntry<Session>;
+};
+
+type ComponentProps = ComponentOnyxProps & {
+    someProp: string;
+};
+
+function Component({session, someProp}: ComponentProps) {
+    const {windowWidth, windowHeight} = useWindowDimensions();
+    const {translate} = useLocalize();
+    // component's code
+}
+
+// There is no hook alternative for withOnyx yet.
+export default withOnyx<ComponentProps, ComponentOnyxProps>({
+    session: {
+        key: ONYXKEYS.SESSION,
+    },
+})(Component);
 ```
 
+### Stateless components vs Pure Components vs Class based components vs Render Props - When to use what?
+
+Class components are DEPRECATED. Use function components and React hooks.
+
+[https://react.dev/reference/react/Component#migrating-a-component-with-lifecycle-methods-from-a-class-to-a-function](https://react.dev/reference/react/Component#migrating-a-component-with-lifecycle-methods-from-a-class-to-a-function)
+
+### Composition 
+
+Avoid the usage of `compose` function to compose HOCs in TypeScript files. Use nesting instead.
+
+  > Why? `compose` function doesn't work well with TypeScript when dealing with several HOCs being used in a component, many times resulting in wrong types and errors. Instead, nesting can be used to allow a seamless use of multiple HOCs and result in a correct return type of the compoment. Also, you can use [hooks instead of HOCs](#hooks-instead-of-hocs) whenever possible to minimize or even remove the need of HOCs in the component.
```

Comment:
```suggestion
> Why? `compose` function doesn't work well with TypeScript when dealing with several HOCs being used in a component, many times resulting in wrong types and errors. Instead, nesting can be used to allow a seamless use of multiple HOCs and result in a correct return type of the compoment. Also, you can use [hooks instead of HOCs](#hooks-instead-of-hocs) whenever possible to minimize or even remove the need of HOCs in the component.
```